### PR TITLE
Add feature importance. Fixes #111

### LINKED ISSUE #111: Feature importance
A figure of feature importances. Can probably be adapted from an old commit. 

Would love to see proper documentation of chosen metric.

### PR DESCRIPTION
**Pull request by HLasse**
Adds feature importance plot and table for both sklearn/xgboost and EBM models.

### CODE REVIEW

File: src/psycopt2d/config/evaluation/default_evaluation.yaml [hunk 1]
```diff
@@ -1 +1,2 @@
-threshold_percentiles: [95, 97, 98, 99, 99.5, 99.7, 99.9]
\ No newline at end of file
+threshold_percentiles: [95, 97, 98, 99, 99.5, 99.7, 99.9]
+top_n_feature_importances: 10
```
Comment (KennethEnevoldsen) [inline src/psycopt2d/config/evaluation/default_evaluation.yaml:2]:
```suggestion
top_n_feature_importances: 10 # top n features to create the plot based on. We also log a table containing all features.
```

File: src/psycopt2d/evaluate_model.py [hunk 2]
```diff
@@ -67,6 +72,32 @@ def evaluate_model(
     # Figures
     plots = {}
 
+    # Feature importance
+    # Check if model has feature_importances_ attribute
+    feature_importances = getattr(pipe["model"], "feature_importances_", None)
+
+    if feature_importances is not None:
+        # Handle EBM and other models a bit differently
```
Comment (KennethEnevoldsen) [inline src/psycopt2d/evaluate_model.py:80]:
```suggestion
        # Handle EBM differently as it autogenerated interaction terms
```

File: src/psycopt2d/visualization/feature_importance.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    sorted_idx = feature_importances.argsort()[::-1]
+
+    x_values = np.array(column_names)[sorted_idx][:top_n_feature_importances]
+    y_values = feature_importances[sorted_idx][:top_n_feature_importances]
+
+    return plot_bar_chart(
+        x_values=x_values,
+        y_values=y_values,
+        x_title="Feature name",
+        y_title="Feature importance",
```
Comment (KennethEnevoldsen) [inline src/psycopt2d/visualization/feature_importance.py:40]: Also add the actual metric name

**Comment by HLasse** [comment]
@KennethEnevoldsen fixed changes. Accept so we can merge?

**Comment by HLasse** [comment]
@KennethEnevoldsen !

File: src/psycopt2d/visualization/base_charts.py [hunk 4]
```diff
@@ -9,21 +9,26 @@ def plot_bar_chart(
...
+        x_values (Iterable): The x values of the bar chart.
+        y_values (Iterable): The y values of the bar chart.
+        x_title (str): title of x axis
+        y_title (str): title of y axis
+        sort_x (Optional[Iterable[int]], optional): order of values on the x-axis. Defaults to None.
+        sort_y (Optional[Iterable[int]], optional): order of values on the y-axis. Defaults to None.
 
     Returns:
-        alt.Chart: An altair chart object.
+        alt.Chart: _description_
```
Comment (KennethEnevoldsen) [inline src/psycopt2d/visualization/base_charts.py:26]: Fix docstring

File: src/psycopt2d/evaluate_model.py [hunk 5]
```diff
@@ -21,8 +24,10 @@
 
 def evaluate_model(
     cfg,
+    pipe: Pipeline,
     eval_dataset: pd.DataFrame,
     y_col_name: str,
+    train_col_names: Iterable[str],
```
Comment (KennethEnevoldsen) [inline src/psycopt2d/evaluate_model.py:30]: missing docstring

File: src/psycopt2d/visualization/feature_importance.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+    Sklearn's standard feature importance metric is "gain"/information gain,
+    which is the decrease in node impurity after the dataset is split on an
+    attribute. Node impurity is measured by Gini impurity and is maximal when
+    the distribution of the two classes in a node is even, and minimal when the
+    classes are perfectly split.
+
+    Args:
+        val_X_column_names (Iterable[str]): Feature names
+        feature_importances (np.ndarray): Feature importances
```
Comment (KennethEnevoldsen) [inline src/psycopt2d/visualization/feature_importance.py:24]: Error in arg name and missing argument description.

**Review by KennethEnevoldsen — CHANGES_REQUESTED** [review]

**Comment by KennethEnevoldsen** [comment]
@HLasse, you can also use the re-request the review instead of @'ing

**Comment by HLasse** [comment]
@KennethEnevoldsen I prefer the passive-aggressive @ <3

File: src/psycopt2d/evaluate_model.py [hunk 7]
```diff
@@ -21,11 +24,32 @@
...
+    7. F1 by time until diagnosis
+
+    Args:
+        cfg (_type_): The hydra config from the run
+        pipe (Pipeline): Pipeline including the model
+        eval_dataset (pd.DataFrame): Evalaution split
+        y_col_name (str): Label column name
+        train_col_names (Iterable[str]): Column names for all predictors
+        y_hat_prob_col_name (str): Column name containing pred_proba output
+        run [wandb.run]: WandB run
```
Comment (KennethEnevoldsen) [inline src/psycopt2d/evaluate_model.py:51]:
```suggestion
        run (wandb.run): WandB run
```

File: src/psycopt2d/evaluate_model.py [hunk 8]
```diff
@@ -21,11 +24,32 @@
...
+    1. AUC
+    2. Table of performance by pred_proba threshold
+    3. Feature importance
+    4. Sensitivity by time to outcome
+    5. AUC by calendar time
+    6. AUC by time from first visit
+    7. F1 by time until diagnosis
+
+    Args:
+        cfg (_type_): The hydra config from the run
```
Comment (KennethEnevoldsen) [inline src/psycopt2d/evaluate_model.py:45]: missing type

**Review by KennethEnevoldsen — CHANGES_REQUESTED** [review]
There were two errors in documentation. They should be a quick fix.

**Comment by MartinBernstorff** [comment]
Thanks for the review, @KennethEnevoldsen! If the changes are minor and uncontronversial (like these), I think it'd be good for us to just implement them as a reviewer. This'd remove one review cycle with little added cost. 

Let's try it and see how it goes :-)

**Review by MartinBernstorff — APPROVED** [review]

**Comment by MartinBernstorff** [comment]
(When we "Request changes", there's no way for others to override. I'd love to merge the branch right now, but as far as I can see, there's no way for me to override your request 😉)

**Review by KennethEnevoldsen — APPROVED** [review]
